atoma-p2p/src/service.rs

```diff
@@ -0,0 +1,1138 @@
+    ///
+    /// - Uses cryptographic signatures to prevent impersonation
+    /// - Relies on the state manager's authoritative record of node ownership
+    /// - Protects against node ID spoofing attacks
+    #[instrument(level = "debug", skip_all)]
+    pub async fn verify_node_small_id_ownership(
+        node_small_id: u64,
+        signature: &[u8],
+        state_manager_sender: Sender<StateManagerEvent>,
+    ) -> Result<(), AtomaP2pNodeError> {
+        let signature = Signature::from_bytes(signature).map_err(|e| {
+            error!("Failed to parse signature");
+            AtomaP2pNodeError::SignatureParseError(e.to_string())
+        })?;
+        let public_key_bytes = signature.public_key_bytes();
+        let public_key =
+            PublicKey::try_from_bytes(signature.scheme(), public_key_bytes).map_err(|e| {
+                error!("Failed to extract public key from bytes, with error: {e}");
+                AtomaP2pNodeError::SignatureParseError(e.to_string())
+            })?;
+        let sui_address = SuiAddress::from(&public_key);
+        let (sender, receiver) = oneshot::channel();
+        if let Err(e) = state_manager_sender.send((
+            AtomaP2pEvent::VerifyNodeSmallIdOwnership {
+                node_small_id,
+                sui_address: sui_address.to_string(),
+            },
+            Some(sender),
+        )) {
+            error!(
+                target = "atoma-p2p",
+                event = "failed_to_send_event_to_state_manager",
+                error = %e,
+                "Failed to send event to state manager"
+            );
+            return Err(AtomaP2pNodeError::StateManagerError(e));
+        }
+        match receiver.await {
+            Ok(result) => {
+                if result {
+                    Ok(())
+                } else {
+                    Err(AtomaP2pNodeError::NodeSmallIdOwnershipVerificationError(
+                        "Node small ID ownership verification failed".to_string(),
+                    ))
+                }
+            }
+            Err(e) => {
+                error!("Failed to receive result from state manager, with error: {e}");
+                Err(AtomaP2pNodeError::NodeSmallIdOwnershipVerificationError(
+                    e.to_string(),
+                ))
+            }
+        }
+    }
+
+    /// Validates the messages received from the P2P network
+    ///
+    /// This function validates the sharing node public addresses messages received from the P2P network by checking the signature and the node small ID ownership.
+    ///
+    /// # Arguments
+    ///
+    /// * `response` - The message received from the P2P network
+    /// * `state_manager_sender` - The sender of the state manager
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if the message is valid, or a `AtomaP2pNodeError` if any step fails.
+    #[instrument(level = "debug", skip_all)]
+    pub async fn validate_usage_metrics_message(
```

Comment:
Let's an integration test for this given it's a combination of all of the functions